File: src/gateway_v2/buyer.ts
import { assertNoPayloadFields, clientMethods } from "./client_contracts";
import type { OpaqueId, SharedRequestMetadata } from "./client_contracts";

export const verifyProviderClientMethod = clientMethods.verify_provider;
export const requestAccessClientMethod = clientMethods.request_access;
export const estimateCostClientMethod = clientMethods.estimate_cost;
export const createBillingSessionClientMethod = clientMethods.create_billing_session;

export interface BuyerBudgetCap {
  perSessionCents: number;
  lifetimeCents: number;
  lifetimeSpendCents?: number;
  currency?: string;
}

export interface BuyerPolicyEnvelope {
  agentId: OpaqueId;
  delegatedUserId: OpaqueId;
  delegatedAccountId: OpaqueId;
  scopes: string[];
  budgetCap: BuyerBudgetCap;
  humanApprovalRequired?: boolean;
  humanApprovalState?: "not_required" | "pending" | "approved" | "denied";
}

export interface VerifyProviderRequest {
  metadata: SharedRequestMetadata;
  sellerId: OpaqueId;
  listingId: OpaqueId;
  policyEnvelope: BuyerPolicyEnvelope;
  sellerVerificationRef: string;
  provenanceAttestationRef: string;
  termsUseRightsRef: string;
  qualityProfileRef: string;
  sampleReceiptRef: string;
}

export interface VerifyProviderResponse {
  verificationId: OpaqueId;
  sellerId: OpaqueId;
  listingId: OpaqueId;
  status: string;
  checkedRefs: string[];
  payloadBytesReturned: 0;
}

export interface RequestAccessRequest {
  metadata: SharedRequestMetadata;
  buyerAccountId: OpaqueId;
  sellerId: OpaqueId;
  listingId: OpaqueId;
  requestedUse: string;
  policyEnvelope: BuyerPolicyEnvelope;
}

export interface AccessRequestResponse {
  accessRequestId: OpaqueId;
  buyerAccountId: OpaqueId;
  sellerId: OpaqueId;
  listingId: OpaqueId;
  state: "pending_seller_approval" | "approved" | "denied";
  approvalWorkflowRef: string;
  paymentStateCreated: boolean;
  invocationStateCreated: boolean;
}

export interface EstimateCostRequest {
  metadata: SharedRequestMetadata;
  buyerAccountId: OpaqueId;
  listingId: OpaqueId;
  requestedUnits?: number;
  policyEnvelope: BuyerPolicyEnvelope;
}

export interface CostEstimateResponse {
  estimateId: OpaqueId;
  listingId: OpaqueId;
  projectedAmountCents: number;
  platformFeeBps: number;
  platformFeeCents: number;
  sellerAmountCents: number;
  currency: string;
  nonBinding: true;
  quoteRequiredForBilling: true;
}

export interface CreateBillingSessionRequest {
  metadata: SharedRequestMetadata;
  buyerAccountId: OpaqueId;
  sellerId: OpaqueId;
  quoteId: OpaqueId;
  accessRequestId: OpaqueId;
  termsAcceptanceId: OpaqueId;
  budgetCapCents: number;
  paymentState: "authorized" | "completed" | "payment_required" | "payment_failed";
  policyEnvelope: BuyerPolicyEnvelope;
}

export interface BillingSessionResponse {
  billingSessionId: OpaqueId;
  quoteId: OpaqueId;
  accessRequestId: OpaqueId;
  termsAcceptanceId: OpaqueId;
  buyerAccountId: OpaqueId;
  state: "authorized" | "payment_required" | "payment_failed";
  amountCents: number;
  currency: string;
  accessGranted: false;
}

export interface GatewayBuyerClientOptions {
  baseUrl: string;
  apiKey?: string;
  fetch?: typeof fetch;
}

export class GatewayBuyerClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: GatewayBuyerClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.fetchImpl = options.fetch ?? fetch;
  }

  VerifyProvider(request: VerifyProviderRequest): Promise<VerifyProviderResponse> {
    return this.post("/v1/gateway/buyer/verify_provider", request, false) as Promise<VerifyProviderResponse>;
  }

  RequestAccess(request: RequestAccessRequest): Promise<AccessRequestResponse> {
    return this.post("/v1/gateway/buyer/request_access", request, true) as Promise<AccessRequestResponse>;
  }

  EstimateCost(request: EstimateCostRequest): Promise<CostEstimateResponse> {
    return this.post("/v1/gateway/buyer/estimate_cost", request, false) as Promise<CostEstimateResponse>;
  }

  CreateBillingSession(request: CreateBillingSessionRequest): Promise<BillingSessionResponse> {
    return this.post("/v1/gateway/buyer/billing_sessions", request, true) as Promise<BillingSessionResponse>;
  }

  verifyProvider(request: VerifyProviderRequest): Promise<VerifyProviderResponse> {
    return this.VerifyProvider(request);
  }

  requestAccess(request: RequestAccessRequest): Promise<AccessRequestResponse> {
    return this.RequestAccess(request);
  }

  estimateCost(request: EstimateCostRequest): Promise<CostEstimateResponse> {
    return this.EstimateCost(request);
  }

  createBillingSession(request: CreateBillingSessionRequest): Promise<BillingSessionResponse> {
    return this.CreateBillingSession(request);
  }

  private async post(path: string, request: { metadata: SharedRequestMetadata }, idempotent: boolean): Promise<unknown> {
    assertNoPayloadFields(Object.keys(request));
    const headers: Record<string, string> = {
      "content-type": "application/json",
      "x-request-id": request.metadata.requestId,
    };
    if (idempotent && request.metadata.idempotencyKey) {
      headers["idempotency-key"] = request.metadata.idempotencyKey;
    }
    if (this.apiKey) {
      headers.authorization = `Bearer ${this.apiKey}`;
    }

    const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
      method: "POST",
      headers,
      body: JSON.stringify(request),
    });
    if (!response.ok) {
      throw new Error(`gateway buyer request failed: ${response.status} ${response.statusText}`);
    }
    return response.json();
  }
}

File: python/aim_node/gateway_v2/buyer.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

import httpx

from .contracts import CLIENT_METHODS, GatewaySurface, reject_payload_contract_fields


VERIFY_PROVIDER_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.VERIFY_PROVIDER]
REQUEST_ACCESS_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.REQUEST_ACCESS]
ESTIMATE_COST_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.ESTIMATE_COST]
CREATE_BILLING_SESSION_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.CREATE_BILLING_SESSION]


@dataclass(frozen=True)
class BuyerBudgetCap:
    per_session_cents: int
    lifetime_cents: int
    lifetime_spend_cents: int = 0
    currency: str = "USD"


@dataclass(frozen=True)
class BuyerPolicyEnvelope:
    agent_id: str
    delegated_user_id: str
    delegated_account_id: str
    scopes: list[str]
    budget_cap: BuyerBudgetCap
    human_approval_required: bool = False
    human_approval_state: str = "not_required"


@dataclass(frozen=True)
class VerifyProviderRequest:
    metadata: dict[str, Any]
    seller_id: str
    listing_id: str
    policy_envelope: BuyerPolicyEnvelope
    seller_verification_ref: str
    provenance_attestation_ref: str
    terms_use_rights_ref: str
    quality_profile_ref: str
    sample_receipt_ref: str


@dataclass(frozen=True)
class RequestAccessRequest:
    metadata: dict[str, Any]
    buyer_account_id: str
    seller_id: str
    listing_id: str
    requested_use: str
    policy_envelope: BuyerPolicyEnvelope


@dataclass(frozen=True)
class EstimateCostRequest:
    metadata: dict[str, Any]
    buyer_account_id: str
    listing_id: str
    policy_envelope: BuyerPolicyEnvelope
    requested_units: int = 1


@dataclass(frozen=True)
class CreateBillingSessionRequest:
    metadata: dict[str, Any]
    buyer_account_id: str
    seller_id: str
    quote_id: str
    access_request_id: str
    terms_acceptance_id: str
    budget_cap_cents: int
    payment_state: str
    policy_envelope: BuyerPolicyEnvelope


class GatewayBuyerClient:
    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        client: httpx.Client | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.client = client or httpx.Client()

    def VerifyProvider(self, request: VerifyProviderRequest) -> dict[str, Any]:
        return self._post("/v1/gateway/buyer/verify_provider", request, idempotent=False)

    def RequestAccess(self, request: RequestAccessRequest) -> dict[str, Any]:
        return self._post("/v1/gateway/buyer/request_access", request, idempotent=True)

    def EstimateCost(self, request: EstimateCostRequest) -> dict[str, Any]:
        return self._post("/v1/gateway/buyer/estimate_cost", request, idempotent=False)

    def CreateBillingSession(self, request: CreateBillingSessionRequest) -> dict[str, Any]:
        return self._post("/v1/gateway/buyer/billing_sessions", request, idempotent=True)

    def verify_provider(self, request: VerifyProviderRequest) -> dict[str, Any]:
        return self.VerifyProvider(request)

    def request_access(self, request: RequestAccessRequest) -> dict[str, Any]:
        return self.RequestAccess(request)

    def estimate_cost(self, request: EstimateCostRequest) -> dict[str, Any]:
        return self.EstimateCost(request)

    def create_billing_session(self, request: CreateBillingSessionRequest) -> dict[str, Any]:
        return self.CreateBillingSession(request)

    def _post(self, path: str, request: Any, *, idempotent: bool) -> dict[str, Any]:
        data = _drop_none(asdict(request))
        reject_payload_contract_fields(list(data.keys()))
        metadata = data["metadata"]
        headers = {
            "content-type": "application/json",
            "x-request-id": str(_pick(metadata, "requestId", "request_id")),
        }
        idempotency_key = _pick(metadata, "idempotencyKey", "idempotency_key")
        if idempotent and idempotency_key:
            headers["idempotency-key"] = str(idempotency_key)
        if self.api_key:
            headers["authorization"] = f"Bearer {self.api_key}"

        response = self.client.post(f"{self.base_url}{path}", json=data, headers=headers)
        response.raise_for_status()
        return response.json()


def _pick(data: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        if key in data:
            return data[key]
    return None


def _drop_none(data: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in data.items() if value is not None}
